### ISSUE TITLE
Expose logstash index settings as spec settings

### ISSUE THREAD

**Issue by mrdavidlaing** [logsearch/logsearch-boshrelease#40]
The [logstash index settings](https://github.com/logsearch/logsearch-boshrelease/blob/master/jobs/elasticsearch/templates/config/templates/logstash_default.json.erb), eg:

```
            "number_of_shards" : 4,
        "number_of_replicas" : 1,
```

should be configurable per deployment

**Comment by dpb587** [comment]
I'd like to change this issue a bit and propose...

First, we generalize [`logstash_default.json.erb`](https://github.com/logsearch/logsearch-boshrelease/blob/master/jobs/elasticsearch/templates/config/templates/logstash_default.json.erb) since it still has a few of our specifics in it, like `message_data`. Additionally, we should probably get rid of the `_ttl` default in favor of the runtime-configured interval.

Then, because Elasticsearch supports [template merging](http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/indices-templates.html#multiple-templates) we should figure out how/if bosh should manage those user-override templates.
- We could do an inline YAML string value, but managing a JSON template in a YAML file sounds like an unfortunate idea. Is it possible for manifests to embed or reference a file elsewhere (without requiring `spiff` or something tricky)?
- We could just say that you need to manage customized templates yourself (like we're currently doing) which just involves manual `curl -XPUT -d @my-template.json` http://192.0.2.1:9200/_template/my-template` calls when you change the template.

**Comment by aristotelesneto** [comment]
BOSH manifests can leverage in-line ruby code - I generally tend to use that instead of spiff when I'm creating releases, for some dynamism in instance sizing, like so:

```
<%
num_ingestor = 2
num_elasticsearch = 5

pool_size=num_ingestor+num_elasticsearch
%>
```

Then further down on the manifest I can do:

```
...
resource_pools:
- name: common
  size: <%= pool_size %>
...
jobs:
- name: ingestor
  instances: <%= num_ingestor %>
  networks:
  - name: sys
    static_ips:
    <% num_ingestor.times do |i| %>
    <%= "- 172.31.201.#{i+41}\n" %>
    <% end %>
...
```

That could possibly be of use if you wanted to read an external json file into the manifest.

Is there any particular reason why those two values are index dependent (logstash-\* only), instead of a global elasticsearch config? Do we expect to have different index configurations on the same ES cluster?

**Comment by dpb587** [comment]
Woah, I didn't realize manifests were ERB-like. That could be immensely useful...

Regarding `logstash-*`, I don't expect to have different index configurations, but the rest of the config is rather logstash-specific already, so it seems reasonable to keep it all logstash-specific. Additionally, some people might cheat and reuse the cluster for non-logstash things (e.g. [marvel](http://www.elasticsearch.org/overview/marvel/)) where other defaults may be more appropriate. The committed template is intended to be a very general starting point. In most cases, people will probably want to customize the template to their own needs if/when they start getting fancy with their data, and at that point they can match indices however they like.

**Comment by dpb587** [comment]
I think I'm of the opinion that BOSH should not manage the templates after all. Instead, the user should manage them in whatever way works best for them (most likely a post-step in their BOSH deployment workflow). I have the following reasons...
- these templates affect persistent data and reverting a deployment to a prior configuration with different templates will not revert how elasticsearch utilizes that existing data (i.e. it seems very different than a property which defines how much memory to use)
- a change to the template as a BOSH property would require a BOSH restart of at least a master node (whereas raw API calls outside of BOSH deploys can avoid restarts altogether)
- I'm pretty sure it's more fragile to manage templates on the filesystem because during a deployment there may be discrepancies on which index template gets used (old vs new) depending on which elasticsearch node gets the indexing request if it's a new index

Instead, I think we should
1. change the current [template](https://github.com/logsearch/logsearch-boshrelease/blob/6228d133033c184d8582d6e59c68507ba5bfbd60/jobs/elasticsearch/templates/config/templates/logstash_default.json.erb) to be a very, very low priority (rather than the current 10)
2. further generalize the current template, removing our custom fields and bringing it back up to date
3. provide a helper script that users can reuse which accepts two arguments - a directory and a `host:ip` - and iterates through `*.json` files in the directory, `PUT`ing them to `host:ip/_template/*` and deleting any templates (aside from the default logsearch one and marvel one) which no longer exist on the file system.

---

@wdneto, does this seem like a valid approach with respect to BOSH-isms?

**Comment by aristotelesneto** [comment]
Upon reading your comment, I get the impression that there are two very distinct types of operations that need to be carried out on the production clusters you've been dealing with:
1. Initial deployment / Re-deployment of a cluster, with its stock standard settings
2. Additional modifications to a running cluster (which are not necessarily becoming default for all clusters) required for other purposes (or potential purposes, eg marvel).

From a BOSH management perspective, unless you're doing number 1, I'd say the operation should be done by way of BOSH Errands - which essentially means BOSH will spin up a new VM, completely independent from the deployment, and execute arbitrary commands (eg. API calls, template imports, etc). Worth noting that an errand does not have direct access to the deployed machines.

One of the things that's help me use BOSH more effectively is to keep in mind the [cattle paradigm](https://blog.engineyard.com/2014/pets-vs-cattle) - whereby state is held mostly outside the machines, and they are reconfigured from a template each time they need re-deploying. Because of that, I'd personally vouch for any approach thats more leaning towards the generic need of logsearch as a whole, than individually.

(this thought is what led me to originally suggest that those two settings should be global rather than specific to logstash)

**Comment by aristotelesneto** [comment]
It's worth noting that what I'm saying might be completely crazy and irrelevant, since I'm unfamiliar with the challenges you're currently facing.

**Comment by mrdavidlaing** [comment]
@dpb587, Since the settings in question are very much about the data you are adding to the cluster (rather than the cluster infrastructure itself), they seem to belong with the `logsearch-filters-???` along with the parsing rules for the incoming data and the kibana dashboards for viewing the parsed data.

What do you think about sticking the common settings inside `logsearch-filters-common`, with the expectation (and tooling) that if you want to customise them you will do so alongside your filters and dashboards in `logsearch-filters-my_custom_filters`?

**Comment by dpb587** [comment]
Okay, I've been convinced that an errand would be valuable and correct way of managing templates for this sort of thing. This change will provide the following benefits:
- unifying how templates/mappings are deployed to the cluster
- allows both the filters and their respective mappings via index templates to be versioned in the same way
- provides a built-in, automation method for deploying them for us to document and encourage
- with this consistency we can move forward with other ideas on deployments sharing their mappings

Here's a rough spec for whenever this becomes a priority or a motivated contributor comes along and picks this up...

---

First, add a property (probably `elasticsearch.templates`), whose value is an array. The array values should be a hash with a single key-value pair with the key being the template name and value being the template contents (json). Ordering is unimportant. Example:

```
properties:
  elasticsearch:
    templates:
      - logstash-cityindex: <%= embedded file %>
      - logsearch-shipper-mappings: <%= embedded file %>
      - app-mapping: <%= embedded file %>
```

Second, create the BOSH errand. It will also need to be configured with an elasticsearch API endpoint that it can communicate with (probably reuse `elasticsearch.host`). The run script should simply iterate through all those key value pairs and execute a curl command like:

```
cat <<EOF | curl -X PUT <%= p('elasticsearch.host') %>:9200/_template/<%= key %> -d
<%= value %>
EOF
```

If the value of the pair is empty, the template should be deleted instead:

```
curl -X DELETE <%= p('elasticsearch.host') %>:9200/_template/<%= key %>
```

Assuming all curl commands are successful, the errand was a success. If there's a failure, it should probably exit immediately with the expectation that the system is in an unknown state and must be manually investigated, recovered, and re-run the errand.

This particular change should _not_ attempt to manipulate existing index mappings. Those shall stay a manual operation, if needed. This is a feature meant only for indices which can be dynamically created.

Elasticsearch has a documentation page about [index templates](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html). bosh.io has a documentation page about [errands](http://bosh.io/docs/jobs.html).

**Comment by dpb587** [comment]
This was resolved by PR #144
